File: server/dive_utils/serializers/frame_metadata.py
import csv
from dataclasses import dataclass
import io
import os
import re
from typing import Dict, Iterable, List, Mapping, Optional, Sequence, Tuple

from dive_utils import constants
from dive_utils.serializers import viame

FRAME_METADATA_SOURCE_EXTENSIONS = {'.txt', '.csv'}


@dataclass(frozen=True)
class ParsedFrameMetadata:
    source_name: Optional[str]
    header: List[str]
    rows: List[Dict[str, str]]
    join_columns: List[str]
    payload_columns: List[str]
    records: Dict[str, Dict[str, str]]


def normalize_key(value: str) -> str:
    """Normalize a media filename the same way valid_image_names_dict keys images."""
    basename = os.path.basename(str(value).strip())
    stem, ext = os.path.splitext(basename)
    if ext.lower().lstrip('.') in constants.allValidLargeImageFormats:
        return stem
    return basename


def parse_table(text: str) -> Tuple[List[str], List[Dict[str, str]]]:
    return _parse_table_rows(_read_rows(text))


def _parse_table_rows(raw_rows: List[List[str]]) -> Tuple[List[str], List[Dict[str, str]]]:
    if not raw_rows:
        return [], []

    header = [cell.strip() for cell in raw_rows[0]]
    if not all(header):
        return [], []

    rows: List[Dict[str, str]] = []
    for raw_row in raw_rows[1:]:
        values = [cell.strip() for cell in raw_row]
        if not any(values):
            continue
        values = values[: len(header)] + [''] * max(0, len(header) - len(values))
        rows.append(dict(zip(header, values)))
    return header, rows


def find_join_columns(
    header: Sequence[str],
    rows: Iterable[Mapping[str, str]],
    media_keys: Mapping[str, int],
) -> List[str]:
    return _find_join_columns_for_keys(header, rows, _normalized_media_keys(media_keys))


def _find_join_columns_for_keys(
    header: Sequence[str],
    rows: Iterable[Mapping[str, str]],
    normalized_media_keys: set,
) -> List[str]:
    materialized_rows = list(rows)
    return [
        column
        for column in header
        if any(
            row.get(column) and normalize_key(row[column]) in normalized_media_keys
            for row in materialized_rows
        )
    ]


def is_frame_metadata(text: str, media_keys: Mapping[str, int]) -> bool:
    return parse_frame_metadata_source(text, media_keys) is not None


def parse_frame_metadata_source(
    text: str,
    media_keys: Mapping[str, int],
    source_name: Optional[str] = None,
) -> Optional[ParsedFrameMetadata]:
    raw_rows, delimiter = _read_rows_with_delimiter(text)
    if delimiter == ',' and viame.is_viame_csv_rows(raw_rows):
        return None

    header, rows = _parse_table_rows(raw_rows)
    if not header or not rows:
        return None

    normalized_media_keys = _normalized_media_keys(media_keys)
    join_columns = _find_join_columns_for_keys(header, rows, normalized_media_keys)
    if not join_columns:
        return None

    payload_columns = [column for column in header if column not in join_columns]
    if not payload_columns:
        return None

    records: Dict[str, Dict[str, str]] = {}
    for row in rows:
        for column in join_columns:
            key = normalize_key(row.get(column, ''))
            if key in normalized_media_keys:
                records[key] = {field: row.get(field, '') for field in header}

    if not records:
        return None

    return ParsedFrameMetadata(
        source_name=source_name,
        header=list(header),
        rows=rows,
        join_columns=join_columns,
        payload_columns=payload_columns,
        records=records,
    )


def select_frame_metadata_source(
    candidates: Iterable[Tuple[str, str]],
    media_keys: Mapping[str, int],
) -> Optional[ParsedFrameMetadata]:
    matches: List[ParsedFrameMetadata] = []
    for source_name, text in candidates:
        if not is_frame_metadata_source_name(source_name):
            continue
        source = parse_frame_metadata_source(text, media_keys, source_name=source_name)
        if source is not None:
            matches.append(source)

    if len(matches) != 1:
        return None
    return matches[0]


def _read_rows(text: str) -> List[List[str]]:
    return _read_rows_with_delimiter(text)[0]


def _read_rows_with_delimiter(text: str) -> Tuple[List[List[str]], Optional[str]]:
    first_line = _first_nonempty_line(text)
    if first_line is None:
        return [], None

    delimiter = _sniff_delimiter(first_line)
    if delimiter is None:
        return [re.split(r'\s+', line.strip()) for line in text.splitlines() if line.strip()], None

    reader = csv.reader(io.StringIO(text), delimiter=delimiter)
    return [
        [cell.strip() for cell in row]
        for row in reader
        if row and any(cell.strip() for cell in row)
    ], delimiter


def _first_nonempty_line(text: str) -> Optional[str]:
    for line in text.splitlines():
        if line.strip():
            return line.strip()
    return None


def _sniff_delimiter(line: str) -> Optional[str]:
    if ',' in line:
        return ','
    if '\t' in line:
        return '\t'
    return None


def _normalized_media_keys(media_keys: Mapping[str, int]) -> set:
    return {normalize_key(key) for key in media_keys}


def is_frame_metadata_source_name(source_name: str) -> bool:
    return os.path.splitext(source_name.lower())[1] in FRAME_METADATA_SOURCE_EXTENSIONS

File: client/platform/desktop/backend/serializers/frameMetadata.ts
import parseSync from 'csv-parse/lib/sync';
import path from 'path';

type FrameMetadataRow = Record<string, string>;
type MediaKeys = Map<string, number> | Record<string, number>;

interface ParsedFrameMetadata {
  sourceName?: string;
  header: string[];
  rows: FrameMetadataRow[];
  joinColumns: string[];
  payloadColumns: string[];
  records: Record<string, FrameMetadataRow>;
}

const imageExtensions = new Set([
  'png',
  'jpg',
  'jpeg',
  'sgi',
  'bmp',
  'pgm',
  'nitf',
  'tif',
  'tiff',
  'ntf',
  'vrt',
  'r0',
  'r1',
  'r2',
  'r3',
  'r4',
  'r5',
  'r6',
]);

function normalizeKey(value: string): string {
  const basename = path.basename(String(value).trim());
  const ext = path.extname(basename);
  const extension = ext.toLowerCase().replace(/^\./, '');
  if (imageExtensions.has(extension)) {
    return path.basename(basename, ext);
  }
  return basename;
}

function parseTable(text: string): { header: string[]; rows: FrameMetadataRow[] } {
  const rawRows = readRows(text);
  if (rawRows.length === 0) {
    return { header: [], rows: [] };
  }

  const header = rawRows[0].map((cell) => cell.trim());
  if (!header.every((cell) => cell.length > 0)) {
    return { header: [], rows: [] };
  }

  const rows: FrameMetadataRow[] = [];
  rawRows.slice(1).forEach((rawRow) => {
    const values = rawRow.map((cell) => cell.trim());
    if (!values.some((cell) => cell.length > 0)) {
      return;
    }
    const row: FrameMetadataRow = {};
    header.forEach((field, index) => {
      row[field] = values[index] || '';
    });
    rows.push(row);
  });
  return { header, rows };
}

function findJoinColumns(
  header: string[],
  rows: FrameMetadataRow[],
  mediaKeys: MediaKeys,
): string[] {
  const normalizedMediaKeys = normalizedKeySet(mediaKeys);
  return header.filter((column) => rows.some((row) => (
    row[column] && normalizedMediaKeys.has(normalizeKey(row[column]))
  )));
}

function isFrameMetadata(text: string, mediaKeys: MediaKeys): boolean {
  return parseFrameMetadataSource(text, mediaKeys) !== null;
}

function parseFrameMetadataSource(
  text: string,
  mediaKeys: MediaKeys,
  sourceName?: string,
): ParsedFrameMetadata | null {
  if (isViameCsv(text)) {
    return null;
  }

  const { header, rows } = parseTable(text);
  if (header.length === 0 || rows.length === 0) {
    return null;
  }

  const joinColumns = findJoinColumns(header, rows, mediaKeys);
  if (joinColumns.length === 0) {
    return null;
  }

  const payloadColumns = header.filter((column) => !joinColumns.includes(column));
  if (payloadColumns.length === 0) {
    return null;
  }

  const records: Record<string, FrameMetadataRow> = {};
  const normalizedMediaKeys = normalizedKeySet(mediaKeys);
  rows.forEach((row) => {
    joinColumns.forEach((column) => {
      const key = normalizeKey(row[column] || '');
      if (normalizedMediaKeys.has(key)) {
        const record: FrameMetadataRow = {};
        header.forEach((field) => {
          record[field] = row[field] || '';
        });
        records[key] = record;
      }
    });
  });

  if (Object.keys(records).length === 0) {
    return null;
  }

  return {
    sourceName,
    header,
    rows,
    joinColumns,
    payloadColumns,
    records,
  };
}

function selectFrameMetadataSource(
  candidates: [string, string][],
  mediaKeys: MediaKeys,
): ParsedFrameMetadata | null {
  const matches: ParsedFrameMetadata[] = [];
  candidates.forEach(([sourceName, text]) => {
    if (!isTextCandidate(sourceName)) {
      return;
    }
    const source = parseFrameMetadataSource(text, mediaKeys, sourceName);
    if (source !== null) {
      matches.push(source);
    }
  });

  if (matches.length !== 1) {
    return null;
  }
  return matches[0];
}

function readRows(text: string): string[][] {
  const firstLine = firstNonemptyLine(text);
  if (firstLine === null) {
    return [];
  }

  const delimiter = sniffDelimiter(firstLine);
  if (delimiter === null) {
    return text
      .split(/\r?\n/)
      .filter((line) => line.trim().length > 0)
      .map((line) => line.trim().split(/\s+/));
  }

  return parseDelimited(text, delimiter);
}

function parseDelimited(text: string, delimiter: ',' | '\t'): string[][] {
  return parseSync(text, {
    delimiter,
    relax_column_count: true,
    // Telemetry sidecars routinely contain bare quote characters (e.g. depth
    // 5"); `relax` treats a stray quote in an unquoted field as a literal
    // character instead of throwing, matching Python's lenient csv.reader on
    // the server while still honoring properly quoted fields.
    relax: true,
    skip_empty_lines: true,
  }).map((row: string[]) => row.map((cell) => cell.trim()));
}

function firstNonemptyLine(text: string): string | null {
  const line = text.split(/\r?\n/).find((candidate) => candidate.trim().length > 0);
  return line === undefined ? null : line.trim();
}

function sniffDelimiter(line: string): ',' | '\t' | null {
  if (line.includes(',')) {
    return ',';
  }
  if (line.includes('\t')) {
    return '\t';
  }
  return null;
}

function normalizedKeySet(mediaKeys: MediaKeys): Set<string> {
  if (mediaKeys instanceof Map) {
    return new Set(Array.from(mediaKeys.keys()).map((key) => normalizeKey(key)));
  }
  return new Set(Object.keys(mediaKeys).map((key) => normalizeKey(key)));
}

function isTextCandidate(sourceName: string): boolean {
  return ['.txt', '.csv'].includes(path.extname(sourceName).toLowerCase());
}

function isViameCsv(text: string): boolean {
  const rows = parseDelimited(text, ',');
  let hasHeader = false;
  let hasDataRow = false;
  let firstRowIsDetection = false;
  let seenDataRow = false;

  rows.forEach((row) => {
    if (row.length === 0) {
      return;
    }
    if (row[0].startsWith('#')) {
      hasHeader = hasHeader || row[0].startsWith('# 1: Detection or Track-id');
      return;
    }
    if (!seenDataRow) {
      seenDataRow = true;
      firstRowIsDetection = isViameDataRow(row);
    }
    if (isViameDataRow(row)) {
      hasDataRow = true;
    }
  });

  // A headerless VIAME CSV (no text header) leads with a detection row; a DIVE
  // export carries the comment header. Telemetry leads with a field-name header
  // that is not VIAME-shaped, so it is left for the frame metadata parser.
  return hasDataRow && (hasHeader || firstRowIsDetection);
}

function isViameDataRow(row: string[]): boolean {
  if (row.length < 9) {
    return false;
  }
  const trackId = Number.parseInt(row[0], 10);
  const frame = Number.parseInt(row[2], 10);
  const bounds = row.slice(3, 7).map((value) => Number.parseFloat(value));
  const fishLength = Number.parseFloat(row[8]);
  return (
    Number.isFinite(trackId)
    && Number.isFinite(frame)
    && bounds.every((value) => Number.isFinite(value))
    && Number.isFinite(fishLength)
  );
}

export {
  FrameMetadataRow,
  MediaKeys,
  ParsedFrameMetadata,
  findJoinColumns,
  isFrameMetadata,
  normalizeKey,
  parseFrameMetadataSource,
  parseTable,
  selectFrameMetadataSource,
};
